### PR TITLE
FFF-103: Async delivery of local messages

### PR DESCRIPTION
**Pull request by adlrocha**
This PR partially addresses #103 and implements and addresses the reviews from #188 related to the broadcast interface and the async delivery of local messages.
- It introduces a `RequestBroadcast` function to the `Network` interface that delivers local messages aynchronously.
- Unifies `Broadcast` and `BroadcastSynchronous`  into a single `Broadcast` function.
- Introduces `SimNetwork` interface as an intermediary step to the removal of IDs from gpbft.Participant (coming next)

### Coming next
Two more PRs extracted from #188 with:
- removal of ID from gpbft.Participant
- addition of the message builder pattern

### CODE REVIEW

File: sim/network.go
Comment (masih) [inline sim/network.go:81]:
```suggestion
func (nf *networkFor) Log(format string, args ...any) {
```

**Review by masih — COMMENTED** [review]
LGTM 👍  

This PR would benefit from a review from @anorth as the reviewer of the PR from which changes here are derived from 🙏

File: gpbft/gpbft.go
Comment (Kubuxu) [inline gpbft/gpbft.go:819]: This line should be removed, given that the broadcast is done async.

Comment (adlrocha) [reply gpbft/gpbft.go:819]: Good catch, thanks!

File: sim/network.go
Comment (masih) [reply sim/network.go:81]: @adlrocha how come this is resolved? 😆

Comment (adlrocha) [reply sim/network.go]: Thank you for flagging it, I fixed it somewhere else and forgot to push the one you commented: https://github.com/filecoin-project/go-f3/pull/259/commits/23a55118c16eb19fec6f76da605e352112748495#diff-4fa4e1c1a7b22e16444c976a3fecfaf2d47f9c48ca7d1ebcb955034e9daab511L148

File: sim/adversary/adversary.go
Comment (anorth) [inline sim/adversary/adversary.go:17]: It's now much less obvious at the call sites when the special adversary powers are being invoked. I would prefer this method were named specially when the synchronous behaviour is invoked, and adversaries that aren't using it just use the same method as the normal participant.

File: gpbft/api.go
Comment (anorth) [inline gpbft/api.go:42]: Changing semantics is fine. I don't think we need to rename this method if the old behaviour is gone.

File: sim/network.go
Comment (anorth) [inline sim/network.go:82]: Thanks. Please enable tracing in some test and check that the logs look reasonable. You've added a prefix here but I didn't spot where the existing prefix has been removed from call sites, so it might be duplicated now.

**Review by anorth — COMMENTED** [review]
Not blocking, but:
- please check the logs are good
- I'd prefer the adversary special powers were more obvious when used

File: sim/network.go
Comment (adlrocha) [reply sim/network.go:82]: Good catch, thanks. This was in preparation for the ID removal, but after working in #267 I realized with my proposal is no longer needed, so I will remove it from here.

File: gpbft/api.go
Comment (adlrocha) [reply gpbft/api.go:42]: I went with this approach because naming it `Broadcast` leads to a collision among the interfaces used by `Host`.

File: sim/adversary/adversary.go
Comment (adlrocha) [reply sim/adversary/adversary.go:17]: I see, I went with this approach per this comment: https://github.com/filecoin-project/go-f3/pull/188/files#r1595990496 but this makes sense. I'll revert to how it was before.

**Comment by Kubuxu** [comment]
@anorth could you re-review?

File: sim/adversary/repeat.go
Comment (anorth) [inline sim/adversary/repeat.go:103]: This can just use the normal `RequestBroadcast` method. It doesn't need to pass the sender explicitly.

File: sim/adversary/spam.go
Comment (anorth) [inline sim/adversary/spam.go:74]: Ditto use the vanilla `RequestBroadcast` here.

File: sim/network.go
Comment (anorth) [inline sim/network.go:120]: I think we want to deliver locally here too, with the same logic as `RequestBroadcast`. I don't think we need the sender parameter at all. I think the only difference here is the synchronous mode. You should be able to delegate to this from `RequestBroadcast` too.

File: gpbft/api.go
Comment (anorth) [reply gpbft/api.go:44]: I don't think you need that method in the adversary host, though.

File: sim/adversary/adversary.go
Comment (anorth) [inline sim/adversary/adversary.go:14]: I think this method is unnecessary. All callers use the message sender as the sender parameter. Now that this PR is delivering messages locally, I don't think we need a sender parameter at all, not in BroadcastSynchronous either.

File: sim/network.go
Comment (anorth) [reply sim/network.go:82]: Prefix is still here...

File: gpbft/api.go
Comment (adlrocha) [reply gpbft/api.go:44]: You are right, this is no longer necessary. Thanks!

File: sim/network.go
Comment (adlrocha) [reply sim/network.go:82]: Removed it from this review, as it adds nothing and it will make more sense once we address the review of my followup PRs. Sorry for the unnecessary noise.

**Comment by adlrocha** [comment]
@anorth, thank you for your review. I think I addressed them all. Let me know if there's still something pending.

**Comment by adlrocha** [comment]
@masih, @Kubuxu, actually, I realized that probably we do not need to merge the PRs as a chain because they are not strongly dependent. Let's get this one merged to `main`, and I will point to `main` from #274 and have a final review on that one before merging. Thanks! :pray:

**Review by anorth — APPROVED** [review]
Not sure what happened to your CI tests, but LGTM now. Thanks!

**Comment by anorth** [comment]
Running tests locally:
- In one instance, panic with `fatal error: concurrent map writes`. This is most likely to be sharing of the BLS signing backend between concurrent tests
- In other instances seem to hang spinning forever (seems like what's happening in CI too, with the container then being killed_

**Comment by anorth** [comment]
I think the repeat adversary is now receiving and repeating the messages that it has sent itself. This is because of the change in semantics of requesting broadcast.

**Comment by adlrocha** [comment]
Thank you for the feedback, @anorth. You were right, I think I found it and fixed it in this commit: https://github.com/filecoin-project/go-f3/pull/259/commits/4a6f98ae7b44a5f3883756d4b926af17e3fad9ae. Let's wait for CI see if it is happy.

**Pull request merged**
